src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -331,113 +405,61 @@ InstanceKlass* LoaderConstraintTable::find_constrained_klass(Symbol* name,
   return NULL;
 }
 
-void LoaderConstraintTable::ensure_loader_constraint_capacity(
-                                                     LoaderConstraintEntry *p,
-                                                    int nfree) {
-    if (p->max_loaders() - p->num_loaders() < nfree) {
-        int n = nfree + p->num_loaders();
-        ClassLoaderData** new_loaders = NEW_C_HEAP_ARRAY(ClassLoaderData*, n, mtClass);
-        memcpy(new_loaders, p->loaders(), sizeof(ClassLoaderData*) * p->num_loaders());
-        p->set_max_loaders(n);
-        FREE_C_HEAP_ARRAY(ClassLoaderData*, p->loaders());
-        p->set_loaders(new_loaders);
-    }
-}
-
-
-void LoaderConstraintTable::extend_loader_constraint(LoaderConstraintEntry* p,
-                                                     Handle loader,
+void LoaderConstraintTable::merge_loader_constraints(Symbol* class_name,
+                                                     LoaderConstraint* p1,
+                                                     LoaderConstraint* p2,
                                                      InstanceKlass* klass) {
-  ensure_loader_constraint_capacity(p, 1);
-  int num = p->num_loaders();
-  p->set_loader(num, loader());
-  p->set_num_loaders(num + 1);
-  LogTarget(Info, class, loader, constraints) lt;
-  if (lt.is_enabled()) {
-    ResourceMark rm;
-    lt.print("extending constraint for name %s by adding loader[%d]: %s %s",
-               p->name()->as_C_string(),
-               num,
-               ClassLoaderData::class_loader_data(loader())->loader_name_and_id(),
-               (p->klass() == NULL ? " and setting class object" : "")
-               );
-  }
-  if (p->klass() == NULL) {
-    p->set_klass(klass);
-  } else {
-    assert(klass == NULL || p->klass() == klass, "constraints corrupted");
-  }
-}
 
+  // Copy into the shorter of the constraints.
+  LoaderConstraint* dest = p1->num_loaders() <= p2->num_loaders() ? p1 : p2;
+  LoaderConstraint* src = dest == p1 ? p2 : p1;
```

Comment:
Is the purpose to reduce the number of copies? If so, you should copy into the one with more loaders.

Reply:
Yes, I should have done that.